lastseen.py

```diff
@@ -0,0 +1,87 @@
+from itertools import chain
+from datetime import datetime
+from errbot import BotPlugin, botcmd
+
+USAGE_STR = 'Nick not specified. (usage: !last_seen <nick>)'
+CONFIG_TEMPLATE = {
+    # locale-preffered format is default
+    'TIME_FORMAT': '%c',
+    'MAX_CHARS': 40
+}
+
+
+class LastSeen(BotPlugin):
+    def configure(self, configuration):
+        if configuration is not None and configuration != {}:
+            config = dict(chain(CONFIG_TEMPLATE.items(),
+                                configuration.items()))
+        else:
+            config = CONFIG_TEMPLATE
+        super(LastSeen, self).configure(config)
+
+    def get_configuration_template(self):
+        return CONFIG_TEMPLATE
+
+    def activate(self):
+        super(LastSeen, self).activate()
+        try:
+            if type(self['last_seen']) is not dict:
+                self['last_seen'] = {}
+        except KeyError:
+            self['last_seen'] = {}
+
+    def callback_message(self, msg):
+        user = msg.nick
+        last_seens = self['last_seen']
+
+        last_seens[user] = {}
+        last_seens[user]['time'] = datetime.now()
+        last_seens[user]['msg'] = msg.body
+
+        self['last_seen'] = last_seens
+
+    def get_msg(self, timedelta, user, text):
+        hours = timedelta.seconds // 3600
+        minutes = (timedelta.seconds - (hours * 3600)) // 60
+        min_format = 'minute' if minutes == 1 else 'minutes'
+
+        if hours > 0:
+            hour_format = 'hour' if hours == 1 else 'hours'
+            params = [user, hours, hour_format, minutes, min_format, text]
+            msg = '{} was last seen {} {} and {} {} ago, saying "{}".'
+            return msg.format(*params)
+
+        params = [user, minutes, min_format, text]
+        msg = '{} was last seen {} {} ago, saying "{}".'
+        return msg.format(*params)
+
+    @botcmd(split_args_with=' ')
+    def last_seen(self, msg, args):
+        """Return last activity and its time of given user.
+
+        Examples:
+            !last_seen <user>
+        """
+        last_seens = self['last_seen']
+
+        if len(args) < 1:
+            return USAGE_STR
+        user = args[0]
+
+        try:
```

Comment:
@JakubNvk I believe this try block is a bit too big. My first reading of line 87 was that `text` may not be defined. Furthermore, I think this approach exploits the `KeyError` exception a bit, as `last_seens[user]['time']` would raise this exception even if `last_seens[user]` existed but `last_seens[user]['time']` did not (though that basically won't happen unless someone tampers with the DB).

Could we maybe replace the exception with a simple check, something along the lines of:

```python
if not user in last_seens:
    return 'We have not seen {} yet.'.format(user)
```

After that line, we can be reasonably sure that `last_seens[user]['time']` and `last_seens[user]['msg']` exist and thus the code can be made a bit more readable.